packages/toolpad-components/src/DataGrid.tsx

```diff
@@ -134,6 +146,33 @@ const DataGridComponent = React.forwardRef(function DataGridComponent(
     columnsInitRef.current = true;
   }, [hasColumnsDefined, rows, nodeRuntime]);
 
+  React.useEffect(() => {
+    if (!nodeRuntime || rows.length <= 0 || idFieldInitRef.current) {
+      return;
+    }
+
+    if (rowIdFieldProp || Boolean(rows?.[0]?.id)) {
+      idFieldInitRef.current = true;
+      return;
+    }
+
+    const mappedRows = addIdField(rows);
+    nodeRuntime.setProp('rows', mappedRows);
```

Comment:
There seems to be a misunderstanding on what this method is for. I'm updating the name https://github.com/mui/mui-toolpad/pull/342. Hopefully that makes it more clear